### PR TITLE
Tcp mode

### PR DESCRIPTION
**Pull request by WillSewell**
@zimbatm do you mind reviewing this?

I will squash the commits into a single one, but I am keeping them separate for now, just in case.

### CODE REVIEW

File: example.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
+
+var Stagger = require('./lib/stagger');
+
+var stagger = Stagger.getDefault({
+  "pid": "1234",
+  "cmd": "./hello"
```
Comment (zimbatm) [inline example.js]: getDefault is a singleton so it can't be configurable, it's used in various places inside of our code. The library should find out the current pid and command from the process by default.

File: example.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+});
+
+// function is called when report is requested
+stagger.registerCount('count0', function() {
+  return 1;
+});
+
+function go() {
+  setTimeout(function () {
```
Comment (zimbatm) [inline example.js]: It's better to use setInterval in JavaScript to avoid drifting in time unless that's what you want. Eg: when you want to have back-pressure.

File: lib/protocol/constants.js [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    "report_all": 0x30,
+    "register_process": 0x41,
+    "stats_partial": 0x42,
+    "stats_complete": 0x43
+  }
+};
+
+module.exports = {
+  "magicBytes": 0x8384,
+  "version": 0x00,
```
Comment (zimbatm) [inline lib/protocol/constants.js:25]: The version could be made part of the magic bytes for now.

File: lib/protocol/encoder.js [hunk 4]
```diff
@@ -0,0 +1,20 @@
+var msgpack = require('msgpack2');
```
Comment (zimbatm) [inline lib/protocol/encoder.js:1]: I think msgpack5 is now the latest message-pack library. Not a requirement though since we don't use binary data.

File: lib/protocol/parser.js [hunk 5]
```diff
@@ -0,0 +1,69 @@
+var msgpack = require('msgpack2');
+
+var constants = require('./constants');
+
+// Streaming data parser
+function Parser(callback, errback) {
```
Comment (zimbatm) [inline lib/protocol/parser.js]: Callback and errback are usually used for promises where you can only have a single callback XOR errback associated as a return value. The parser gets some data and emits either messages or errors so an EventEmitter might make more sense.

File: lib/protocol/parser.js [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+  this.rawData = new Buffer(0);
+}
+
+var prototype = Parser.prototype;
+
+prototype.feedData = function(buffer) {
+  this.rawData = Buffer.concat([this.rawData, buffer]);
+
+  if (this.rawData.length >= 8) {
+    var msgLength = 8 + this.rawData.readUInt32BE(4);
```
Comment (zimbatm) [inline lib/protocol/parser.js:21]: 4 is the byte position in the buffer yes ?

File: example.js (hunk 1 above)
Comment (zimbatm) [inline example.js]: `process.pid` and `process.argv[0]` are part of the node globals

File: lib/protocol/parser.js (hunk 6 above)
Comment (WillSewell) [inline lib/protocol/parser.js:21]: Yes

File: lib/stagger/tcp_client.js [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+          break;
+        case 'pair:pong':
+          break; // NOOP
+        default:
+          this.emit("message", method, body);
+          break;
+      }
+    }.bind(this),
+    function(err) {
+      console.log("Parse error: " + err);
```
Comment (zimbatm) [inline lib/stagger/tcp_client.js]: It feels like debug info, the client should probably emit an error as well

File: lib/stagger/tcp_client.js [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+  );
+
+  this.connect(tags, connectionOpts);
+}
+
+util.inherits(TcpClient, events.EventEmitter);
+var prototype = TcpClient.prototype;
+
+prototype.send = function(method, body) {
+  this.socket.write(encode(method, body));
```
Comment (zimbatm) [inline lib/stagger/tcp_client.js]: what happens if the socket is not connected ?

File: example.js (hunk 1 above)
Comment (WillSewell) [inline example.js]: So the library should set all the tags? Should there be a way of the client application to set these too?

File: lib/protocol/constants.js (hunk 3 above)
Comment (WillSewell) [inline lib/protocol/constants.js:25]: I'm tempted to leave it like this just for neater diffs in the future..

File: lib/stagger/tcp_client.js [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+};
+
+// private
+
+prototype.connect = function(tags, connectionOpts) {
```
Comment (zimbatm) [inline lib/stagger/tcp_client.js]: why are tags and connectionOpts passed around instead of being stored on the TcpClient object ?

File: lib/stagger/tcp_client.js [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+
+prototype.pong = function() {
+  this.send("pair:pong", {});
+};
+
+prototype.resetActivity = function() {
+  this.activityAt = Date.now();
+};
+
+prototype.setupActivityCheck = function() {
```
Comment (zimbatm) [inline lib/stagger/tcp_client.js]: Maybe the activity checks could be a wrapper around the TCPConnection object. Right now there is quite a bit of logic in that object.

**Comment by zimbatm** [comment]
It looks really good overall, I'm happy of how much complexity you managed to remove. The parser and encoder are top-notch.

There are a bit of console.log left in various places. As a library it's best not to log directly. Either ask for a logger object or emit relevant events that the user can decide to log.

The TCP connection is the only part where I'm not 100% sure of the behaviour because mostly I'm not sure who calls which methods and in what conditions. I'm not sure if it's possible to reduce the state machine though. Could you annotate all the functions that are supposed to be used internally vs externally and in which conditions ?

File: example.js (hunk 1 above)
Comment (zimbatm) [inline example.js]: Yes if they create their own instance of the stagger.Client. I prefer libraries who don't provide that default singleton but we're kind of stuck with it for the moment.

File: lib/protocol/constants.js (hunk 3 above)
Comment (zimbatm) [inline lib/protocol/constants.js:25]: Yeah no problem

File: lib/protocol/parser.js (hunk 6 above)
Comment (zimbatm) [inline lib/protocol/parser.js:21]: Ok

File: lib/stagger/tcp_client.js (hunk 8 above)
Comment (WillSewell) [inline lib/stagger/tcp_client.js]: Ah good point. I totally missed that.

File: lib/stagger/tcp_client.js (hunk 9 above)
Comment (WillSewell) [inline lib/stagger/tcp_client.js]: Because I'm used to functional languages :P

I don't feel too strongly about it in this case, and want to stick to conventions for a particular language. So I'll make them object members if we think it makes more sense.

File: lib/stagger/tcp_client.js (hunk 10 above)
Comment (WillSewell) [inline lib/stagger/tcp_client.js]: Good idea.

**Comment by WillSewell** [comment]
Thanks for the feedback!

I agree that the TCP connection logic can be hard to follow. I followed the convention of the existing code by having a `// private` comment to mark the point where following methods. Do you think this should be annotated on a per-function basis?

The class opens a TCP connection when it is instantiated. If there is any error on the socket ("close" is emitted), then there will be a 1 second delay before a reconnection attempt. If parsing of data fails, or the activityCheck fails (ping/pong) then a `end()` is called on the socket by the library which will also result in "close" being emitted, and thus a reconnection.

File: lib/stagger/tcp_client.js (hunk 8 above)
Comment (WillSewell) [inline lib/stagger/tcp_client.js]: What do you think the behaviour should be if it's not connected? Just discards the messages, or keep buffering them and send them all once it reconnects (obviously there will have to be some limit on the number that are buffered).

File: lib/stagger/tcp_client.js (hunk 10 above)
Comment (WillSewell) [inline lib/stagger/tcp_client.js]: Looking at it now I can't see a good way of separating these: the activity checker needs methods from TCPClient: like shutting down the socket, and TCPClient client needs to access methods from the activity checker: e.g. resetActivity().

**Comment by WillSewell** [comment]
Regarding taking a logger as a parameter: if the client of the library calls `getDefault` (i.e. no logger), should the library still log, or not at all?

File: lib/stagger/tcp_client.js (hunk 8 above)
Comment (mdpye) [inline lib/stagger/tcp_client.js]: Late metrics are discarded by stagger, so there's no point queuing them

File: lib/stagger/tcp_client.js (hunk 10 above)
Comment (mdpye) [inline lib/stagger/tcp_client.js]: See an example (using emitted events, how appropriate for node!) https://github.com/pusher/em-hiredis/blob/master/lib/em-hiredis/support/inactivity_checker.rb
